<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-надежности-заемщиков" data-toc-modified-id="Исследование-надежности-заемщиков-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование надежности заемщиков</a></span><ul class="toc-item"><li><span><a href="#Изучение-общей-информации-о-данных" data-toc-modified-id="Изучение-общей-информации-о-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение общей информации о данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Обработка аномальных значений</a></span><ul class="toc-item"><li><span><a href="#Столбец-children" data-toc-modified-id="Столбец-children-1.2.1.1"><span class="toc-item-num">1.2.1.1&nbsp;&nbsp;</span>Столбец <code>children</code></a></span></li><li><span><a href="#Столбец-days_employed" data-toc-modified-id="Столбец-days_employed-1.2.1.2"><span class="toc-item-num">1.2.1.2&nbsp;&nbsp;</span>Столбец <code>days_employed</code></a></span></li><li><span><a href="#Столбец-gender" data-toc-modified-id="Столбец-gender-1.2.1.3"><span class="toc-item-num">1.2.1.3&nbsp;&nbsp;</span>Столбец <code>gender</code></a></span></li><li><span><a href="#Вычисление-максимального-рабочего-стажа" data-toc-modified-id="Вычисление-максимального-рабочего-стажа-1.2.1.4"><span class="toc-item-num">1.2.1.4&nbsp;&nbsp;</span>Вычисление максимального рабочего стажа</a></span></li></ul></li><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Удаление пропусков</a></span><ul class="toc-item"><li><span><a href="#Столбец-days_employed" data-toc-modified-id="Столбец-days_employed-1.2.2.1"><span class="toc-item-num">1.2.2.1&nbsp;&nbsp;</span>Столбец <code>days_employed</code></a></span></li><li><span><a href="#Столбец-total_income" data-toc-modified-id="Столбец-total_income-1.2.2.2"><span class="toc-item-num">1.2.2.2&nbsp;&nbsp;</span>Столбец <code>total_income</code></a></span></li></ul></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li></ul></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-1.5.4"><span class="toc-item-num">1.5.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков

**Заказчик** — кредитный отдел банка.

**Задача:**
Разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Изучение общей информации о данных

In [1]:
# Импорт библиотек
import pandas as pd
import numpy as np
from pymystem3 import Mystem

import warnings

In [2]:
# Считывание csv-файла
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [3]:
# Задаем количесвто столбцов, которые будут выводиться при обращении к датафреймам
pd.set_option('display.max_columns', None)

# Вывести из под комментария перед финальным запуском:
# Игнорируем предупреждения о возможных изменениях работы функций в будущих версиях в Pandas
warnings.filterwarnings("ignore")

In [4]:
# Выведем первые 20 строчек датафрейма
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [5]:
# Выведем основную информацию о датафрейме
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [6]:
# Проверим информацию о количественных признаках
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод:**
* Названия столбцов корректные, соответствуют "змеиному стилю"
* Присутствуют два столбца содержащих пропущенные значения
* Типы столбцов соответствуют информации хранящихся в них
* в столбцах `days_employed`, `children`, `dob_years` содержатся ненормальные для них значения

## Предобработка данных

### Обработка аномальных значений

#### Столбец `children`

В данных встречаются артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке.  

Явные артефакты присутствуют в столбце `children`. Выведем перечень уникальных значений данного столбца

In [7]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Отрицательное значение детей - скорее всего техническая ошибка, которую следует исправить заменой. Более подробные размышления на эту тему будут приведены при обработке столбца `days_employed`. Строки со значением детей равным 20 - удалим.

In [8]:
data['children'].replace(-1,1,inplace = True)
data = data[data['children'] != 20]

Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.

In [9]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

#### Столбец `days_employed`

 Еще одним артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`.

На 2020 год пенсионный возраст женщин 55, мужчин 60. Заметим, что человек проживший 100 лет и работающий с самого своего рождения может максимально иметь ~36500 дней рабочего стажа(что весьма сомнительно), в столбце ответственном за хранение стажа преимущественно отрицательные значения или значение превосходящие человеческие возможности. Проверим есть ли в данном столбце значения большие 0, но при этом не превосходящие человеческие возможности(36500 дней стажа)

In [10]:
data[(data['days_employed'] >= 0) & (data['days_employed'] < 36500)]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

Такие объекты отсутствуют. Напрашивается два вывода: либо при выгрузке в документ данных произошла техническая ошибка из-за которой корректные данные стали отрицательными или при сборе информации были введены неверные значения. Отсюда можно прийти к трем решениям данной проблемы:

1. Откинуть минус и считать данные, которые являются отрицательными истинными, но содержавшими ошибку артефакт в виде минуса. А значения не укладывающиеся в человеческие возможности привести к максимальному стажу **(выбран данный вариант)**
2. И сключить из выборки путем удаления столбец `days_employed` (т.к. данный столбец не влияет на поставленную задачу, а именно определить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок).
3. Можно попробовать привести все некорректные значения столбца `days_employed` к максимальному стажу опираясь на пол и возраст. Но это может сильно испортить данные

#### Столбец `gender`

Для получения максимального трудового стажа потребуется пол. Проверим все ли с ним в порядке

In [11]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

В поле `gender` мы можем увидеть 3 уникальных гендера ('F', 'M', 'XNA')

Отдельно рассмотрим что за гендер XNA

In [12]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Гендер 'XNA' встречается 1 раз в объекте с индексом 10701(исходной таблицы). С ним можно поступить двумя способами:

1. Не учитывать его вовсе и удалить из данных (т.к. он 1 из 21 тысячи значений)
2. Оставить его **(выбран данный способ)**

#### Вычисление максимального рабочего стажа

Напишем функцию, которая вычисляет максимальный рабочий стаж.

Для этого сначала выведем все уникальные значения количества лет.

In [13]:
dob_years = data['dob_years'].unique()
dob_years

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [14]:
# Отсортируем в порядке возрастания
dob_years.sort()
dob_years

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

В случае, если стаж больше возможного функция изменяет его на максимальный. Так же можно было  заметить, что у присутствуют нулевые значения в возрасте, в 0 лет точно ни на какую работу не возьмут, поэтому заменим все 0 значения на 15 лет, так как минимальное после 0 в наших данных это 19 лет, мы всего лишь сдвинем минимальный возраст на более разумную границу

In [15]:
data['dob_years'].replace(0, 15, inplace = True)

In [16]:
# Выведем минимальное значение лет
data['dob_years'].min()

15

In [17]:
def max_seniority(df):
    '''
    Ищет максимальный трудовой стаж.
    Если стаж больше максимального, то
    функция возвращает максимально возможный стаж,
    основываясь на возраст человека и пол, в противном случае
    возвращает исходный трудовой стаж
    '''

    current_seniority = df['days_employed']
    if current_seniority == np.nan:
        return current_seniority
    else:
        if df['gender'] == 'M' and df['dob_years'] > 60 :
            max_years = 60
        elif df['gender'] == 'F' and df['dob_years'] > 55:
            max_years = 55
        else:
            max_years = df['dob_years']

        max_seniority = (max_years - 14) * 365
        if current_seniority > max_seniority:
            return max_seniority
        elif current_seniority < 0:
            return (current_seniority * -1)
        else:
            return current_seniority

In [18]:
data['days_employed'] = data.apply(max_seniority, axis = 1)

In [19]:
# Проверим изменения
data['days_employed'].describe()

count    19284.000000
mean      4568.293976
std       5238.370823
min         24.141633
25%        926.203547
50%       2190.359362
75%       5513.781905
max      18388.949901
Name: days_employed, dtype: float64

**Вывод:**  
* В данных присутствовали аномальные значения, которые были устранены
* Таблица нуждается в очистке и исправлениях, которую частично удалось провести

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца.

In [20]:
data.isna().sum()

children               0
days_employed       2165
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2165
purpose                0
dtype: int64

#### Столбец `days_employed`

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Выведем пример данных с пропусками в данном столбце

In [21]:
data[data['days_employed'].isnull()].sample(10)

children  days_employed  dob_years education  education_id  \
18349         1            NaN         30    высшее             0   
11711         1            NaN         43   среднее             1   
5112          1            NaN         23    высшее             0   
15091         0            NaN         58   среднее             1   
9558          2            NaN         38   среднее             1   
19542         2            NaN         26   среднее             1   
11782         0            NaN         50   среднее             1   
15622         0            NaN         32    высшее             0   
20966         0            NaN         50   среднее             1   
14910         0            NaN         20   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
18349        женат / замужем                 0      F  госслужащий     0   
11711        женат / замужем                 0      F    сотрудник     0   
5112         женат / замужем                 0      F    сотрудник     0   
15091       гражданский брак                 1      M    сотрудник     0   
9558         женат / замужем                 0      M    компаньон     0   
19542        женат / замужем                 0      F    сотрудник     0   
11782              в разводе                 3      M    сотрудник     0   
15622       гражданский брак                 1      F    сотрудник     0   
20966        женат / замужем                 0      M    сотрудник     0   
14910  Не женат / не замужем                 4      M    сотрудник     0   

       total_income                                 purpose  
18349           NaN                 покупка жилья для семьи  
11711           NaN                             образование  
5112            NaN                 покупка жилья для сдачи  
15091           NaN                   на проведение свадьбы  
9558            NaN  строительство собственной недвижимости  
19542           NaN                 покупка жилья для сдачи  
11782           NaN   получение дополнительного образования  
15622           NaN                                 свадьба  
20966           NaN                                   жилье  
14910           NaN              покупка жилой недвижимости

Заметим, что там, где пропуск в `days_employed`, у того же объекта пропуск и в `total_income`. Пропусков 2174, что достаточно много для того что бы не обращать на них внимания и удалить их. Поэтому логично будет сгруппировать по полу и возрасту, а затем для каждой группы найти медиану и присвоить ее пропускам из той же группы

In [22]:
pivot_days_employed = data.pivot_table(index=['dob_years'],
                                       columns = 'gender',
                                       values = 'days_employed',
                                       aggfunc='median')

Теперь переберем пропуски в стаже работы и заполним их в соответствии со сводной таблицей. Напишем для этого специальную функцию

In [23]:
def fillna_days_employed(df):
    """
    Возвращает медианное значение трудового стажа
    опираясь на пол и возраст, если натыкается на NaN.
    В противном случае возвращает переданное значение обратно
    """
    gender = df['gender']
    age = df['dob_years']
    if pd.isnull(df['days_employed']):
        return pivot_days_employed.loc[age,gender]
    else:
        return df['days_employed']

In [24]:
data['days_employed'] = data.apply(fillna_days_employed, axis = 1)

#### Столбец `total_income`

На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`.

In [25]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

 Убедимся, что все пропуски заполнены.

In [26]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Вывод:**  
* Любые данные не идеальны и могут содержать пропуски
* Пропуски нужно заполнять основываясь на зависящие признаки, так данные получаются более точными

###  Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.

In [27]:
data['total_income'] = data['total_income'].astype(int)

**Вывод:**
1. Зачастую данные в таблице содержаться не в подходящем типе данных, но в данном случае в целом все верно
2. Для более высокой точности стоит привести целочисленные значения к типу *int*, а не оставлять во *float*

### Обработка дубликатов

Сначала приведем все текстовые значения к нижнему регистру, исключая гендер, т.к. он больше не понадобится.

Проанализируем столбцы, в которых должны храниться только строковые значения

In [28]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Сразу можно заметить, что здесь присутствуют дубликаты отличающиеся по регистру. Исправим это

In [29]:
data['education'] = data['education'].str.lower()

In [30]:
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Теперь все выглядит аккуратнее и без дубликатов. Сделаем то же самое и для оставшихся столбцов

In [31]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Тут все гораздо лучше, но все равно следует привести их к одному регистру

In [32]:
data['family_status'] = data['family_status'].str.lower()

Теперь перейдем к столбцу `income_type`

In [33]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Тут и без обработки все в порядке.

Найдем общее количество дубликатов

In [34]:
data.duplicated().sum()

71

Избавимся от них c восстановлением индексов

In [35]:
data  = data.drop_duplicates().reset_index(drop = True)

Проверим оставшееся количество дубликатов

In [36]:
data.duplicated().sum()

0

**Вывод:**  
* Были найдены дубликаты по регистру
* Так же были найдены группы, отличавшиеся регистром букв в названии

## Лемматизация

Ознакомимся со всеми уникальными значениями столбца `purpose`.

In [37]:
data['purpose'].value_counts()

свадьба                                   790
на проведение свадьбы                     764
сыграть свадьбу                           761
операции с недвижимостью                  674
покупка коммерческой недвижимости         658
покупка жилья для сдачи                   650
операции с жильем                         648
операции с коммерческой недвижимостью     646
покупка жилья                             643
жилье                                     642
покупка жилья для семьи                   637
недвижимость                              632
строительство собственной недвижимости    629
операции со своей недвижимостью           626
строительство жилой недвижимости          623
покупка своего жилья                      620
строительство недвижимости                619
покупка недвижимости                      619
ремонт жилью                              605
покупка жилой недвижимости                604
на покупку своего автомобиля              505
заняться высшим образованием      

Вариаций целей получения кредита очень много, некоторые из них содержат орфографические ошибки, но их можно разделить на 4 категории - автомобиль, свадьба, образование и недвижимость. Лемматизируем значения `purpose` с помощью библиотеки `pymystem3`, а после добавим новый столбец `purpose_category` с категориями целей.

In [38]:
m = Mystem()

def find_category(row):
    '''
    Возвращает категорию
    в зависимости от указанной цели кредита
    '''
    lemmas = m.lemmatize(row)
    if 'автомобиль' in lemmas:
        return 'автомобиль'
    if 'свадьба' in lemmas:
        return 'свадьба'
    if 'образование'in lemmas:
        return 'образование'
    return 'недвижимость'

In [39]:
data['purpose_category'] = data['purpose'].apply(find_category)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   14235.000000         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category  
0               покупка жилья     недвижимость  
1     приобретение автомобиля       автомобиль  
2               покупка жилья     недвижимость  
3  дополнительное образование      образование  
4             сыграть свадьбу          свадьба  
5               покупка жилья     недвижимость  
6           операции с жильем     недвижимость  
7                 образование      образование  
8       на проведение свадьбы          свадьба  
9     покупка жилья для семьи     недвижимость

**Вывод:**  
* Все цели получения кредита выделили в 4 группы - так будет намного удобнее анализировать данные.

##  Категоризация данных

С помощью метода `qcut()` разозьем данные в столбце `total_income` на квантили.

In [40]:
income_cut = pd.qcut(data['total_income'], 4)
income_cut.value_counts()

(107528.75, 142594.0]     5461
(20666.999, 107528.75]    5345
(195795.5, 2265604.0]     5345
(142594.0, 195795.5]      5227
Name: total_income, dtype: int64

Для ответа на вопрос о связи дохода клиента с вероятностью возникновения задолженности, разделим клиентов на 4 категории по уровню дохода с помощью функции `find_income_group`.

* `low` - доход < 107528 тыс.р
* `medium` - доход >= 107528 тыс.р и < 142594 тыс.р
* `high` - доход >= 142594 тыс.р и < 195795.5 тыс.р
* `high+` - доход >= 195795.5 тыс.р

In [41]:
def find_income_group(income):
    '''
    Возвращает категорию дохода
    в зависимости от размера дохода
    '''
    if income < 107528:
        return('low')
    if 107528 <= income <= 142594:
        return('medium')
    if 142594 <= income <= 195795.5:
        return('high')
    return('high+')

**Вывод:**
* Разделили клиентов банка на 4 категории по уровню дохода.

## Проверка гипотез

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [42]:
# Создадим сводную таблицу
debt_and_children = data[['children', 'debt']]

In [43]:
clients_with_children = debt_and_children.loc[debt_and_children['children'] > 0]
debtors_with_children = clients_with_children.loc[clients_with_children['debt'] == 1]
debtors_with_children_mean = len(debtors_with_children) / len(clients_with_children)

clients_without_children = debt_and_children.loc[debt_and_children['children'] == 0]
debtors_without_children = clients_without_children.loc[clients_without_children['debt'] == 1]
debtors_without_children_mean = len(debtors_without_children) / len(clients_without_children)

In [44]:
if debtors_with_children_mean > debtors_without_children_mean:
    print('Доля должников с детьми - {:.1%}, это больше, чем доля должников без детей.'.format(debtors_with_children_mean))
else:
    print('Доля должников без детей - {:.1%}, это больше, чем доля должников с детьми.'.format(debtors_without_children_mean))

Доля должников с детьми - 9.2%, это больше, чем доля должников без детей.


**Вывод:**
* Доля должников с детьми больше доли бездетных должников. Отсюда вывод, что наличие детей негативно влияет на возврат кредита в срок.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [45]:
# Создадим сводную таблицу
debtors_family_status = data[['family_status', 'debt']]

debtors_family_status_grouped = debtors_family_status.groupby('family_status').count()
debtors_family_status_grouped['number'] = debtors_family_status.loc[debtors_family_status['debt'] == 1].groupby('family_status').sum()
debtors_family_status_grouped['mean, %'] = debtors_family_status_grouped['number'] / debtors_family_status_grouped['debt'] * 100
debtors_family_status_grouped.sort_values('mean, %', ascending=False)

debt  number   mean, %
family_status                                 
не женат / не замужем   2801     273  9.746519
гражданский брак        4139     385  9.301764
женат / замужем        12290     928  7.550854
в разводе               1193      84  7.041073
вдовец / вдова           955      63  6.596859

**Вывод:**
* Результат исследования показывает, что в группе клиентов со статусом не женат / не замужем доля должников больше всего - почти 9,8%. Меньше всего должников среди вдовцов - 6,6%.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [46]:
# Создадим сводную таблицу
debtors_income = data[['total_income', 'debt']]

debtors_income = debtors_income.loc[debtors_income['debt'] == 1]
debtors_income['income_group'] = debtors_income['total_income'].apply(find_income_group)
debtors_income[['income_group', 'debt']].groupby('income_group').sum().sort_values('debt', ascending = False)

debt
income_group      
medium         480
high           445
low            427
high+          381

**Вывод:**
* Поделили клиентов на 4 категории (равных по количеству клиентов) по уровню дохода. В результате подсчета должников в этих категориях видим, что больше всего таковых в категории дохода `high`. Однако, следует заметить, что количество должников в категориях отличается совсем немного, за исключением категории `high+` - здесь их значительно меньше.

Как результат, можем сказать, что выраженной зависимости между уровнем дохода и возвратом кредита в срок нет.

### Как разные цели кредита влияют на его возврат в срок?

In [47]:
bank_clients_purpose = data.pivot_table(index=['purpose_category', 'purpose'], values='debt', aggfunc='sum')
bank_clients_purpose.sort_values('debt', ascending=False).head(10)

debt
purpose_category purpose                                    
свадьба          свадьба                                  64
                 на проведение свадьбы                    61
                 сыграть свадьбу                          58
недвижимость     операции с недвижимостью                 55
                 строительство недвижимости               54
                 операции с коммерческой недвижимостью    52
                 покупка жилья для сдачи                  52
образование      получение дополнительного образования    51
автомобиль       сделка с подержанным автомобилем         51
                 сделка с автомобилем                     50

Основываясь на результатах этой таблицы, чаще всего должниками становятся в результате взятия кредита на проведение свадьбы и на покупку/строительство недвижимости.

Сгруппируем данные по категории цели кредита и выявим долю должников из каждой категории

In [48]:
# Создадим сводную таблицу
debtors_purpose = data[['purpose_category', 'purpose', 'debt']]

debtors_purpose_grouped = debtors_purpose.groupby('purpose_category').count()
debtors_purpose_grouped['number'] = debtors_purpose.loc[debtors_purpose['debt'] == 1].groupby('purpose_category').sum()
debtors_purpose_grouped['mean, %'] = debtors_purpose_grouped['number'] / debtors_purpose_grouped['debt'] * 100
debtors_purpose_grouped[['mean, %']].sort_values('mean, %', ascending=False)

mean, %
purpose_category          
автомобиль        9.347319
образование       9.229615
свадьба           7.904968
недвижимость      7.238979

**Вывод:**
* В результате группирования по категории цели кредита видим, что количество должников в категориях "автомобиль" и "образование" выделяется в большую сторону. И, хотя при первой проверке сильно выделялись категории "свадьба" и "недвижимость", при использовании относительных величин картина изменилась.

## Общий вывод

В ходе работы искали зависимость между:  
1. Наличием детей и возвратом кредита в срок - должников с детьми больше, чем бездетных, следовательно, **наличие детей негативно влияет на возврат кредита в срок**;
2. Семейным положением и возвратом кредита в срок - зависимость есть, чаще всего становятся должниками люди с семейным положением **не женат / не замужем**;
3. Уровнем дохода и возвратом кредита в срок - выраженная зависимость между уровнем дохода и возвратом кредита в срок **не установлена**;
4. Целью кредита и возвратом кредита в срок - зависимость есть, бОльшая доля должников взяла кредит на **автомобиль или образование**.

Как результат, выявили зависимость между потециальным долгом и семейным положением, наличием детей и целью кредита. Наличие зависимости между уровнем дохода и возвратом кредита в срок не подтверждено.